src/librustc_typeck/check/mod.rs

```diff
@@ -1939,26 +1939,115 @@ fn check_impl_items_against_trait<'tcx>(
+            let snippet = suggestion_signature(&trait_item, tcx);
+            let code = format!("{}{}\n{}", padding, snippet, padding);
+            let msg = format!("implement the missing item: `{}`", snippet);
+            let appl = Applicability::HasPlaceholders;
             if let Some(span) = tcx.hir().span_if_local(trait_item.def_id) {
                 err.span_label(span, format!("`{}` from trait", trait_item.ident));
+                err.tool_only_span_suggestion(sugg_sp, &msg, code, appl);
             } else {
-                err.note_trait_signature(trait_item.ident.to_string(),
-                                         trait_item.signature(tcx));
+                err.span_suggestion_hidden(sugg_sp, &msg, code, appl);
             }
         }
         err.emit();
     }
 
     if !invalidated_items.is_empty() {
         let invalidator = overridden_associated_type.unwrap();
-        span_err!(tcx.sess, invalidator.span, E0399,
-                  "the following trait items need to be reimplemented \
-                   as `{}` was overridden: `{}`",
-                  invalidator.ident,
-                  invalidated_items.iter()
-                                   .map(|name| name.to_string())
-                                   .collect::<Vec<_>>().join("`, `"))
+        span_err!(
+            tcx.sess,
+            invalidator.span,
+            E0399,
+            "the following trait items need to be reimplemented as `{}` was overridden: `{}`",
+            invalidator.ident,
+            invalidated_items.iter()
+                .map(|name| name.to_string())
+                .collect::<Vec<_>>().join("`, `"))
+    }
+}
+
+/// Given a `ty::AssocItem` and a `TyCtxt`, return placeholder code for that associated item.
+/// Similar to `ty::AssocItem::suggestion`, but appropriate for use as the code snippet of a
+/// structured suggestion.
+fn suggestion_signature(assoc: &ty::AssocItem, tcx: TyCtxt<'_>) -> String {
+    match assoc.kind {
+        ty::AssocKind::Method => {
+            // We skip the binder here because the binder would deanonymize all
+            // late-bound regions, and we don't want method signatures to show up
+            // `as for<'r> fn(&'r MyType)`.  Pretty-printing handles late-bound
+            // regions just fine, showing `fn(&MyType)`.
+            let sig = tcx.fn_sig(assoc.def_id);
+            let unsafety = match sig.unsafety() {
+                hir::Unsafety::Unsafe => "unsafe ",
+                _ => "",
+            };
+            let args = sig.inputs()
+                .skip_binder()
+                .iter()
+                .map(|ty| Some(match ty.kind {
+                    ty::Param(param) if param.name == kw::SelfUpper => {
+                        "self".to_string()
+                    }
+                    ty::Ref(reg, ref_ty, mutability) => {
+                        let mutability = match mutability {
+                            hir::Mutability::MutMutable => "mut ",
```

Comment:
Grep for `"mut "` -- there are some cases involving `MutMutable` which could be refactored into a method: `prefix_str`.